DataFormats/Math/interface/alpaka/deltaPhi.h

```
@@ -0,0 +1,35 @@
+#ifndef DataFormats_Math_alpaka_deltaPhi_h
+#define DataFormats_Math_alpaka_deltaPhi_h
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+
+namespace cms::alpakatools {
+
+  // reduce to [-pi,pi]
+  template <typename TAcc, typename T>
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE constexpr T reduceRange(TAcc const& acc, T x) {
```

I was going to comment the `ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE` are not necessary because the function is `constexpr`, but after looking at Alpaka implementation of e.g. `abs()` and `round()` 
https://github.com/alpaka-group/alpaka/blob/a4142d3feb7686d803e1ec5f25d7b2278337f455/include/alpaka/math/Traits.hpp#L864
https://github.com/alpaka-group/alpaka/blob/a4142d3feb7686d803e1ec5f25d7b2278337f455/include/alpaka/math/Traits.hpp#L1331
that are **not** `constexpr`, I'm actually puzzled how this function can be `constexpr`.

---

@makortel when do we expect to switch to C++23? These functions will be constexpr in the standard STL at that point.

---

Indeed, you are right. Compiling and using the functions did not show any errors though, is it that `constexpr` is silently ignored? If that's the case, I guess I should remove it so that it does not cause confusion.

---

> when do we expect to switch to C++23?

Based on the experience with C++20, maybe around 2027.

---

> Indeed, you are right. Compiling and using the functions did not show any errors though, is it that `constexpr` is silently ignored? If that's the case, I guess I should remove it so that it does not cause confusion.

It seems that `constexpr` functions and function templates are allowed to call non-`constexpr` functions as long as there is at least one set of arguments with which the function can be evaluated as a constant expression (and otherwise the program is ill-formed, no diagnostic required...).

From C++20 draft ([N4849](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/n4849.pdf)), 9.2.5.7 (.6 is also relevant)
> If the instantiated template specialization of a constexpr function template or member function of a class template would fail to satisfy the requirements for a constexpr function, that specialization is still a constexpr function, even though a call to such a function cannot appear in a constant expression. If no specialization of the template would satisfy the requirements for a constexpr function when considered as a non-template function, the template is ill-formed, no diagnostic required.

So while our compilers might not warn about the situation, C++20-conforming code would not use `constexpr` here. Although IIUC this constraint was removed from C++23.

---

Since C++23 seems to be quite a bit away for CMSSW, I would go with the C++20 conforming option and remove the `constexpr`.

---

Note that `ALPAKA_FN_INLINE` is much stronger than `inline`, as it tries to force the compiler to actually inline the function.

The regular `inline` should be fine most of the time.

---

> The regular inline should be fine most of the time.

Should I be switching to that in this case? I am not sure whether you were suggesting that I apply this change or it was a more general comment.